contributing/BACKENDS.md

```diff
@@ -1,96 +1,167 @@
+
+```python
+<YOURBACKEND> = '<your_backend>'
+```
+2. **Create Provider Directory**: Establish a new directory at `src/dstack/_internal/core/backends/<YourProvider> `to house your provider’s backend and compute implementations.
+
+
+3. **Backend Implementation:** 
+In `__init__.py`, implement `<YourProvider>Backend`, inheriting from `BaseBackend`. Define the `TYPE` class variable to associate your backend with the newly added enum entry.
+
+4. **Compute Implementation:** 
+In `compute.py`, develop `<YourProvider>Compute`, inheriting from `Compute`.<br> 
+
+You'll need to implement methods like      
+  - `get_offers` It will be called every time the user wants to provision something. Add availability information if possible. 
+  - `run_job` Here you create a compute resource and run `dstack-shim` or `dstack-runner`.
+  - `terminate_instance` This method should not raise an error, if there is no such instance.
+
+5. **Configuration Implementation**:
+- Implement the `<YourProvider>Config` class in `config.py`, inheriting from both `BackendConfig` and `<YourProvider>StoredConfig`. This configuration is accepted by the `<YourProvider>Backend` class.
+
+
+### Configuration Models
+ 1. **Create Configuration Models:**
 
-### Adding VM compute backend
+You may have multiple models for credentials (i.e., default credentials & explicit credentials). 
+ In `src/dstack/_internal/core/models/backends/<YourProvider>.py`, create models for your provider's configuration:
+- `<YourProvider>ConfigInfo:` create a model with all configuration details except credentials.
+- `<YourProvider>ConfigInfoWithCreds`: create a model with credentials.
+- `<YourProvider>ConfigInfoWithCredsPartial`: create a model with all fields optional.
+- `<YourProvider>ConfigValues:` create a model representing UI elements for configurator.
 
-`dstack` expects the following features from your backend:
+2. **Import Models:**
+Ensure all new models are imported into `src/dstack/_internal/core/models/backends/__init__.py`.
+
+### Finalizing Integration
+1. **Implement Configurator:**
+Develop `<YourProvider>Configurator` in `src/dstack/_internal/server/services/backends/configurators/<YourProvider>.py`.
+
+2. **Add YAML Configuration Model:**
+Insert `<YourProvider>Config` in `src/dstack/_internal/server/services/config.py` to represent the provider’s configuration in YAML.
+
+3. **Ensure Safe Import:** 
+Add a safe import for your backend in `src/dstack/_internal/server/services/backends/__init__.py` and update expected backends in tests within `src/tests/_internal/server/routers/test_backends.py.`
+
+
+
+
+
+## Appendix
+### Adding VM Compute Backend
+dstack expects VM backends to have:
 
 - Ubuntu 22.04 LTS
 - Nvidia Drivers 535
 - Docker with Nvidia runtime
 - OpenSSH server
 - External IP & 1 port for SSH (any)
-- cloud-init script (preferable)
+- cloud-init script (preferred)
 - API for creating and terminating instances
+- Safe imports
```

Comment:
I guess my previous comment was not clear, my bad. It should not be there, because it has nothing to do with backends, only Python code in dstack itself.

Reply:
Still unclear, I had safe imports originally under finalising integration this way:
Add a safe import for your backend in `src/dstack/_internal/server/services/backends/__init__.py` 
Should this change?

Reply:
This quote is correct, it should stay. Safe imports are related to this exact file (`src/dstack/_internal/server/services/backends/__init__.py`).